### PR TITLE
flux-start: build system instance test features

### PR DESCRIPTION
**Pull request by garlick**
This PR builds up some test capability for stopping and starting brokers in sharness, to simulate system instance failure modes.  First, it cleans up`flux-start` options, renaming some to add a `--test-` prefix where appropriate, and altering others to mimic similar (better thought out) option naming in the shell:
* `-k,--killer-timeout` is renamed to `--test-exit-timeout` (with original as alias for now)
* `-v,--verbose` can now set a verbosity level (`-v -v`, or `--verbosity=2` for example)
* `--trace-pmi-server` is removed, and PMI tracing is now enabled at verbosity >= 2, like the shell
* `-c,--noclique` is renamed to `--test-pmi-clique=none` (similar to shell pmi.clique option)
* `-D,--rundir` is renamed to `--test-rundir`, and its error handling improved somewhat
* all test options are grouped in the help output, as suggested by @grondo 

One new option was added:
* `--test-exit-mode=leader|any` defines what happens when a broker exits.  The old behavior of starting the exit timer when any broker exits, and exiting with the higest broker exit code, is still the default.  A new mode called `leader` starts the exit timer only when rank 0 exits, and then exits with the rank 0 exit code.

The embedded `flux-start` server added in #3650 gets some new methods: 
* `start.kill` - send a signal to a broker by rank (or to all brokers)
* `start.wait` - get a broker's exit code
* `start.run` - (re-)start a broker

`t/scripts/startctl.py` now has corresponding subcommands for the above.

A new "test personality" called `system` is added that bootstraps flux using a config file instead of PMI, so the brokers can be restarted and be able to reconnect to their peers.

Finally, the `t3300-system-basic.t` sharness script is modified to use the new test personality and restart brokers using `startctl`.

### CODE REVIEW

File: t/sharness.d/flux-sharness.sh
Comment (grondo) [inline t/sharness.d/flux-sharness.sh:94]: Ah, thanks for the extra documentation here!

**Review by grondo — APPROVED** [review]
LGTM, just some minor comments/questions inline that you can feel free to ignore.

File: src/broker/overlay.c
Comment (grondo) [inline src/broker/overlay.c:542]: In the testsuite, for some reason `flux_strerror (errno)` appears to return an empty string here (at least on my system). I have no idea how this could be the case:
```
	flux dmesg|grep broker.debug|grep "overlay child"

2021-06-09T14:38:40.594423Z broker.debug[0]: overlay child 1 62e99b77-5f30-4d67-abb0-06f51e063380 connnected=true Sent hello
2021-06-09T14:38:40.711337Z broker.debug[0]: overlay child 2 01c09cde-bf7f-43d2-b603-45849e4a32f4 connnected=true Sent hello
2021-06-09T14:39:00.358933Z broker.debug[0]: overlay child 2 01c09cde-bf7f-43d2-b603-45849e4a32f4 connnected=false Sent DISCONNECT
2021-06-09T14:39:03.221390Z broker.debug[0]: overlay child 2 3712f6fb-8eb2-416e-b732-6fde0d0e60a9 connnected=true Sent hello
2021-06-09T14:39:05.229186Z broker.debug[0]: overlay child 2 3712f6fb-8eb2-416e-b732-6fde0d0e60a9 connnected=false Mcast failed: 
ok 18 - dump overlay child logs from leader broker
```

Since this log message is only emitted when `errno == EHOSTUNREACH`, perhaps a custom error message could be used here instead?

Edit: this error is only emitted when running the testcase under `FLUX_TEST_VALGRIND=t`. Maybe that is part of the problem.

File: src/cmd/flux-start.c
Comment (garlick) [reply src/cmd/flux-start.c:99]: > Why not take multiple -v supplied on the command line to increase verbosity as we do with other commands?

`-v -v` is handled as equivalent to `--verbose=2` here.  The argument is already optional.

> One drawback of having an option that takes an optional argument is that it can't be combined with other short options

Yeah, that is a bit unfortunate, so `-vv` will fail as a bad argument to `-v`.

**Review by grondo — APPROVED** [review]

File: src/cmd/flux-start.c
Comment (grondo) [reply src/cmd/flux-start.c:99]: > -v -v is handled as equivalent to --verbose=2 here. The argument is already optional.

Yeah, my question is why the difference here compared to all the other flux commands that take multiple `-v` where `-vv` works? Did you need `--verbose=2` explicitly for some reason? It broke for me immediately since I tried `flux start -vvs 4`

File: src/cmd/flux-start.c
Comment (garlick) [reply src/cmd/flux-start.c:200]: Oh thanks, the following change works.  I'm not sure why I didn't do this before - I guess I landed where I did through trial and error.
```diff
diff --git a/src/cmd/flux-start.c b/src/cmd/flux-start.c
index d59825f4c..5b5aaada5 100644
--- a/src/cmd/flux-start.c
+++ b/src/cmd/flux-start.c
@@ -190,13 +190,8 @@ int main (int argc, char *argv[])
     if (optparse_hasopt (ctx.opts, "verbose")) {
         const char *optarg = NULL;
         ctx.verbose = optparse_getopt (ctx.opts, "verbose", &optarg);
-        if (optarg) {
-            char *endptr;
-            errno = 0;
-            ctx.verbose = strtoul (optarg, &endptr, 10);
-            if (errno != 0 || !isdigit (*optarg) || *endptr != '\0')
-                log_msg_exit ("error parsing --verbose option argument");
-        }
+        if (optarg)
+            ctx.verbose = optparse_get_int (ctx.opts, "verbose", 1);
     }
 
     if (optindex < argc) {
```

File: src/broker/overlay.c
Comment (garlick) [reply src/broker/overlay.c:542]: Well that's unexpected!  I went ahead and made the change you suggested.

I thought I had reviewed strerror code and determined that it was safe to use the non-thread-safe version in flux, but of course we do have multiple threads potentially going at it in the broker.  I'll open an issue to revisit that.

File: src/cmd/flux-start.c
Comment (garlick) [reply src/cmd/flux-start.c:99]: Ah, is `-vv` support preferable to `--verbose=2` support?  I leaned the other way for no apparent reason.  Maybe we could have it both ways by registering the short and long options separately?

Comment (grondo) [reply src/cmd/flux-start.c:99]: Yeah, sorry I thought you were aware of that. It is how we use it in `flux-shell` `flux-exec` and `flux mini run`.

Good thought about registering long and short options separately, unfortunately I'm not sure we ever added support for short-option-only options in liboptparse.

You could, however, have a separate option `--loglevel` or something if there is a pressing need to set the verbosity numerically?

Comment (garlick) [reply src/cmd/flux-start.c:99]: I'll go ahead and get rid of the optional argument.  I think you're right that `-vv` is the more common and expected idiom.

Comment (grondo) [reply src/cmd/flux-start.c:99]: perhaps eventually liboptparse could support a flag to make the optional argument apply only to the long option, perhaps using your idea of separating the short from the longopts in the underlying `getopt_long()` configuration.

One other drawback to optional arguments on long opts, however, is that unlike required arguments, the `=` is compulsory if you want to supply an argument, i.e. `--verbose 2` wouldn't work. Perhaps there would be a way to work around this limitation as well.

**Comment by garlick** [comment]
OK, forced a push that was
* rebased on current master to pick up #3706 (optparse improvement for optional arguments)
* simplified `flux start -v,--verbose[=LEVEL]` option handling per above
* addressed the other issue @grondo brought up above
* Added some configurability to the `system` test personality via TEST_UNDER_FLUX_* environment variables (found these were useful while writing tests for a future PR) 
* Moved ipc sockets from "@-space" to the trash directory so more than one `system` personality test can run concurrently (also found while developing another test script)

One thing I forgot to note is that the `system` personality leaves a trash directory behind on success containing the configuration.  This is because the "real" trash directory is inside this one, and it gets cleaned up by the re-execed script.  Since the script is execed  we never get back to the original to clean up.  I'm open to suggestions on how to handle this.  Maybe it's OK for now to leave them behind?

**Comment by garlick** [comment]
Restarted due to job shell input hang (known test issue #3546).

**Comment by grondo** [comment]
> I'm open to suggestions on how to handle this. Maybe it's OK for now to leave them behind?

Probably best to clean it up. This seems to work:

```diff
diff --git a/t/sharness.d/flux-sharness.sh b/t/sharness.d/flux-sharness.sh
index 72adfaddc..8565210e4 100644
--- a/t/sharness.d/flux-sharness.sh
+++ b/t/sharness.d/flux-sharness.sh
@@ -90,6 +90,17 @@ make_bootstrap_config() {
     echo "-o,-Sbroker.quorum=${TEST_UNDER_FLUX_QUORUM:-$full}"
 }
 
+#
+#  Remove any outer trash-directory wrapper used by "system"
+#   personality test_under_flux() tests.
+#
+remove_trashdir_wrapper() {
+    local trashdir=$(dirname $SHARNESS_TRASH_DIRECTORY)
+    case $trashdir in
+        */trash-directory.[!/]*) rm -rf $trashdir
+    esac
+}
+
 #
 #  Reinvoke a test file under a flux instance
 #
@@ -134,6 +145,9 @@ test_under_flux() {
     personality=${2:-full}
     log_file="$TEST_NAME.broker.log"
     if test -n "$TEST_UNDER_FLUX_ACTIVE" ; then
+        if test "$TEST_UNDER_FLUX_PERSONALITY" = "system"; then
+            cleanup remove_trashdir_wrapper
+        fi
         test "$debug" = "t" || cleanup rm "${SHARNESS_TEST_DIRECTORY:-..}/$log_file"
         flux_module_list > module-list.initial
         cleanup check_module_list
@@ -190,6 +204,7 @@ test_under_flux() {
     logopts="-o -Slog-filename=${log_file},-Slog-forward-level=7"
     TEST_UNDER_FLUX_ACTIVE=t \
     TERM=${ORIGINAL_TERM} \
+    TEST_UNDER_FLUX_PERSONALITY="${personality:-default}" \
       exec flux start --test-size=${size} \
                       ${RC1_PATH+-o -Sbroker.rc1_path=${RC1_PATH}} \
                       ${RC3_PATH+-o -Sbroker.rc3_path=${RC3_PATH}} \
```

The `TEST_UNDER_FLUX_PERSONALITY` check is actually optional since the parent of the trash directory will only be removed if it too is named `trash-directory.*`, but it seemed best to have an abundance of caution, and it may be useful to have TEST_UNDER_FLUX_PERSONALITY available elsewhere in the future. (maybe)

**Comment by garlick** [comment]
Thank you!   Just pushed that as a separate commit.  I had not looked far enough to discover the cleanup() function - handy!

**Comment by grondo** [comment]
A couple other aesthetics-only changes for `flux start --help` output

 * Since `flux-start` now has many long options, increase the "option width" to give more room.
 * Add a "doc" header to the test-only section

e.g.
```diff
diff --git a/src/cmd/flux-start.c b/src/cmd/flux-start.c
index 7017a8d7e..367d2a63c 100644
--- a/src/cmd/flux-start.c
+++ b/src/cmd/flux-start.c
@@ -118,6 +118,8 @@ static struct optparse_option opts[] = {
       .usage = "Wrap broker execution in comma-separated arguments"
     },
     /* Option group 2 */
+    { .group = 2,
+       .usage = "\nOptions useful for testing:" },
     { .group = 2,
       .name = "test-size",       .key = 's', .has_arg = 1, .arginfo = "N",
       .usage = "Start a test instance by launching N brokers locally", },
@@ -168,6 +170,9 @@ int main (int argc, char *argv[])
 
     if (!(ctx.opts = optparse_create ("flux-start"))
         || optparse_add_option_table (ctx.opts, opts) != OPTPARSE_SUCCESS
+        || optparse_set (ctx.opts,
+                         OPTPARSE_OPTION_WIDTH,
+                         32) != OPTPARSE_SUCCESS
         || optparse_set (ctx.opts,
                          OPTPARSE_USAGE,
                          usage_msg) != OPTPARSE_SUCCESS)
```

result:
```
Usage: flux-start [OPTIONS] command ...
  -h, --help                    Display this message.
  -v, --verbose=[LEVEL]         Be annoyingly informative by degrees
  -X, --noexec                  Don't execute (useful with -v, --verbose)
  -o, --broker-opts=OPTS        Add comma-separated broker options, e.g. "-o,-v"
      --caliper-profile=PROFILE Enable profiling in brokers using Caliper
                                configuration profile named `PROFILE'
      --wrap=ARGS,...           Wrap broker execution in comma-separated
                                arguments

Options useful for testing:
  -s, --test-size=N             Start a test instance by launching N brokers
                                locally
      --test-hosts=HOSTLIST     Set FLUX_FAKE_HOSTNAME in environment of each
                                broker
      --test-exit-timeout=FSD   After a broker exits, kill other brokers after
                                timeout
      --test-exit-mode=any|leader
                                Trigger exit timer on leader/any broker exit
                                (default=any)
      --test-rundir=DIR         Use DIR as broker run directory
      --test-pmi-clique=single|none
                                Set PMI_process_mapping mode (default=single)

```

Also, this makes me wonder why we don't just suppress the `--caliper-profile` option when flux is not built with Caliper enabled.

If you'd like I can just fix these 3 trivial issues in a follow-on PR.

**Comment by garlick** [comment]
Oh very nice.  I'll go ahead and add those changes here.

**Comment by grondo** [comment]
This LGTM and already has approval if you want to set MWP (and would prefer a separate PR for aesthetic updates to `flux-start --help`)

**Pull request merged**
